CodeEdit/Features/LSP/LSPService.swift

```diff
@@ -0,0 +1,209 @@
+    func documentWasClosed(for languageId: LanguageIdentifier, file fileURL: URL) async throws -> Bool {
+        // TODO: GET FILE TYPE FROM DOCUMENT, USING NEW FILE SOLUTION
+        guard var languageClient = self.languageClient(for: .python) else {
+            logger.error("Failed to get \(languageId.rawValue) client")
+            throw ServerManagerError.languageClientNotFound
+        }
+        return await languageClient.closeDocument(fileURL.absoluteString)
+    }
+
+    /// NOTE: This function is intended to be removed when the frontend is being developed.
+    /// For now this is just for reference of a working example.
+    func testCompletion() async throws {
+        do {
+            guard var languageClient = self.languageClient(for: .python) else {
+                print("Failed to get client")
+                throw ServerManagerError.languageClientNotFound
+            }
+
+            let testFilePathStr = ""
+            let testFileURL = URL(fileURLWithPath: testFilePathStr)
+
+            // Tell server we opened a document
+            _ = await languageClient.addDocument(testFileURL)
+
+            // Completion example
+            let textPosition = Position(line: 32, character: 18)  // Lines and characters start at 0
+            let completions = try await languageClient.requestCompletion(
+                document: testFileURL.absoluteString,
+                position: textPosition
+            )
+            switch completions {
+            case .optionA(let completionItems):
+                // Handle the case where completions is an array of CompletionItem
+                print("\n*******\nCompletion Items:\n*******\n")
+                for item in completionItems {
+                    let textEdits = getCompletionItemEdits(startPosition: textPosition, item: item)
+                    for edit in textEdits {
+                        print(edit)
+                    }
+                }
+
+            case .optionB(let completionList):
+                // Handle the case where completions is a CompletionList
+                print("\n*******\nCompletion Items:\n*******\n")
+                for item in completionList.items {
+                    let textEdits = getCompletionItemEdits(startPosition: textPosition, item: item)
+                    for edit in textEdits {
+                        print(edit)
+                    }
+                }
+
+                print(completionList.items[0])
+
+            case .none:
+                print("No completions found")
+            }
+
+            // Close the document
+            _ = await languageClient.closeDocument(testFilePathStr)
+        } catch {
+            print(error)
+        }
+    }
+
+    /// Attempts to stop a running language server. Throws an error if the server is not found
+    /// or if the language server throws an error while trying to shutdown.
```

Comment:
Is there a way to know which of these two is the case?

Reply:
`server.shutdownAndExit()` throws an error when it fails to shutdown but it's being caught by the do/catch. I'll need to propagate the error upwards so that the UI can handle displaying any errors relating to this.